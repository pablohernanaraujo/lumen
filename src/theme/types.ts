import type { ImageStyle, TextStyle, ViewStyle } from 'react-native';

import type { colorTokens } from './colors';
import type { baseTokens } from './tokens';

export type Mode = keyof typeof colorTokens;

export type BaseTokens = typeof baseTokens;

export type ColorTokens = typeof colorTokens;

export type ThemeColors = (typeof colorTokens)[keyof typeof colorTokens];

export type Theme = BaseTokens & {
  colors: ThemeColors;
  mode: Mode;
};

export type StyleFunction<T> = (theme: Theme) => T;

export type NamedStyles<T> = {
  [P in keyof T]: ViewStyle | TextStyle | ImageStyle;
};

export type StylesOrFunction<T extends NamedStyles<T>> = T | StyleFunction<T>;

export interface ThemeContextValue {
  theme: Theme;
  mode: Mode;
  setMode: (mode: Mode) => void;
  toggleMode: () => void;
}

export interface ThemeProviderProps {
  children: React.ReactNode;
  preference?: Mode;
  persistenceKey?: string;
}
