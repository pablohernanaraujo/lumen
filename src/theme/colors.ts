export const colorTokens = {
  light: {
    background: '#FFFFFF',
    surface: '#F8F9FA',
    surfaceSecondary: '#F1F3F4',
    text: {
      primary: '#1A1A1A',
      secondary: '#6B7280',
      tertiary: '#9CA3AF',
      inverse: '#FFFFFF',
    },
    primary: {
      50: '#EFF6FF',
      100: '#DBEAFE',
      200: '#BFDBFE',
      300: '#93C5FD',
      400: '#60A5FA',
      500: '#3B82F6',
      600: '#2563EB',
      700: '#1D4ED8',
      800: '#1E40AF',
      900: '#1E3A8A',
      main: '#3B82F6',
      dark: '#1D4ED8',
      light: '#93C5FD',
    },
    secondary: {
      50: '#FFF7ED',
      100: '#FFEDD5',
      200: '#FED7AA',
      300: '#FDBA74',
      400: '#FB923C',
      500: '#F97316',
      600: '#EA580C',
      700: '#C2410C',
      800: '#9A3412',
      900: '#7C2D12',
      main: '#F97316',
      dark: '#C2410C',
      light: '#FDBA74',
    },
    success: {
      main: '#10B981',
      dark: '#059669',
      light: '#6EE7B7',
    },
    warning: {
      main: '#F59E0B',
      dark: '#D97706',
      light: '#FCD34D',
    },
    error: {
      main: '#EF4444',
      dark: '#DC2626',
      light: '#FCA5A5',
    },
    info: {
      main: '#3B82F6',
      dark: '#1D4ED8',
      light: '#93C5FD',
    },
    border: '#E5E7EB',
    borderLight: '#F3F4F6',
    borderDark: '#D1D5DB',
    overlay: 'rgba(0, 0, 0, 0.5)',
  },
  dark: {
    background: '#000000',
    surface: '#111111',
    surfaceSecondary: '#1F1F1F',
    text: {
      primary: '#FFFFFF',
      secondary: '#A1A1AA',
      tertiary: '#71717A',
      inverse: '#000000',
    },
    primary: {
      50: '#1E3A8A',
      100: '#1E40AF',
      200: '#1D4ED8',
      300: '#2563EB',
      400: '#3B82F6',
      500: '#60A5FA',
      600: '#93C5FD',
      700: '#BFDBFE',
      800: '#DBEAFE',
      900: '#EFF6FF',
      main: '#60A5FA',
      dark: '#2563EB',
      light: '#BFDBFE',
    },
    secondary: {
      50: '#7C2D12',
      100: '#9A3412',
      200: '#C2410C',
      300: '#EA580C',
      400: '#F97316',
      500: '#FB923C',
      600: '#FDBA74',
      700: '#FED7AA',
      800: '#FFEDD5',
      900: '#FFF7ED',
      main: '#FB923C',
      dark: '#EA580C',
      light: '#FED7AA',
    },
    success: {
      main: '#34D399',
      dark: '#10B981',
      light: '#A7F3D0',
    },
    warning: {
      main: '#FBBF24',
      dark: '#F59E0B',
      light: '#FEF3C7',
    },
    error: {
      main: '#F87171',
      dark: '#EF4444',
      light: '#FECACA',
    },
    info: {
      main: '#60A5FA',
      dark: '#3B82F6',
      light: '#BFDBFE',
    },
    border: '#374151',
    borderLight: '#4B5563',
    borderDark: '#1F2937',
    overlay: 'rgba(0, 0, 0, 0.8)',
  },
} as const;
