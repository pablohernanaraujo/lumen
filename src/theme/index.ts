export { colorTokens } from './colors';
export { makeStyles } from './make-styles';
export { ThemeProvider, useTheme } from './theme-provider';
export { baseTokens } from './tokens';
export type {
  BaseTokens,
  ColorTokens,
  EmphasisLevel,
  Mode,
  NamedStyles,
  StyleFunction,
  StylesOrFunction,
  Theme,
  ThemeColors,
  ThemeContextValue,
  ThemeProviderProps,
} from './types';
export {
  createButtonStyle,
  createCardStyle,
  createInputStyle,
  createShadow,
  createTextStyle,
  getColorOpacity,
  getContrastColor,
  isLightColor,
} from './utils';
