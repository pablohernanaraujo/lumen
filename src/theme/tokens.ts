export const baseTokens = {
  spacing: {
    none: 0,
    xs: 4,
    sm: 8,
    md: 16,
    lg: 24,
    xl: 32,
    xxl: 48,
  },
  radii: {
    sm: 4,
    md: 8,
    lg: 16,
    xl: 24,
    round: 999,
  },
  typography: {
    family: {
      light: 'Nunito-Light',
      regular: 'Nunito-Regular',
      medium: 'Nunito-Medium',
      semibold: 'Nunito-SemiBold',
      bold: 'Nunito-Bold',
      mono: 'Courier',
    },
    size: {
      xs: 12,
      sm: 14,
      md: 16,
      lg: 18,
      xl: 20,
      xxl: 24,
      xxxl: 32,
    },
    weight: {
      light: '300' as const,
      regular: '400' as const,
      medium: '500' as const,
      semibold: '600' as const,
      bold: '700' as const,
    },
    lineHeight: {
      tight: 1.2,
      normal: 1.4,
      relaxed: 1.6,
    },
  },
  shadows: {
    sm: {
      shadowColor: '#000000',
      shadowOffset: {
        width: 0,
        height: 1,
      },
      shadowOpacity: 0.05,
      shadowRadius: 2,
      elevation: 2,
    },
    md: {
      shadowColor: '#000000',
      shadowOffset: {
        width: 0,
        height: 2,
      },
      shadowOpacity: 0.1,
      shadowRadius: 4,
      elevation: 4,
    },
    lg: {
      shadowColor: '#000000',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.15,
      shadowRadius: 8,
      elevation: 8,
    },
  },
  opacity: {
    disabled: 0.6,
    overlay: 0.8,
    subtle: 0.4,
  },
  emphasis: {
    pure: 1,
    high: 0.87,
    medium: 0.6,
    low: 0.38,
  },
} as const;
