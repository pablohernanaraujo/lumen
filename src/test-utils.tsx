import React, { type FC, type ReactElement } from 'react';
import {
  render,
  type RenderOptions,
  type RenderResult,
} from '@testing-library/react-native';

import { ThemeProvider } from './theme';
import type { Mode } from './theme/types';

interface AllTheProvidersProps {
  children: React.ReactNode;
  mode?: Mode;
}

const AllTheProviders: FC<AllTheProvidersProps> = ({
  children,
  mode = 'light',
}) => <ThemeProvider preference={mode}>{children}</ThemeProvider>;

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  mode?: Mode;
  wrapper?: React.ComponentType<{ children: React.ReactNode }>;
}

export const customRender = (
  ui: ReactElement,
  { mode = 'light', wrapper, ...renderOptions }: CustomRenderOptions = {},
): RenderResult => {
  const Wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
    const content = <AllTheProviders mode={mode}>{children}</AllTheProviders>;

    return wrapper
      ? React.createElement(wrapper, { children: content })
      : content;
  };

  return render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
};

// Re-export everything from React Native Testing Library
export * from '@testing-library/react-native';

// Override the default render with our custom one
export { customRender as render };

// Create snapshot matcher helper
export const createSnapshot = (
  component: ReactElement,
  mode: Mode = 'light',
): void => {
  const { toJSON } = customRender(component, { mode });
  expect(toJSON()).toMatchSnapshot();
};

// Create both light and dark mode snapshots
export const createDualSnapshots = (component: ReactElement): void => {
  describe('snapshots', () => {
    it('should render correctly in light mode', () => {
      createSnapshot(component, 'light');
    });

    it('should render correctly in dark mode', () => {
      createSnapshot(component, 'dark');
    });
  });
};

// Export TestWrapper for use in tests
export const TestWrapper: FC<{ children: React.ReactNode; mode?: Mode }> = ({
  children,
  mode = 'light',
}) => <AllTheProviders mode={mode}>{children}</AllTheProviders>;
