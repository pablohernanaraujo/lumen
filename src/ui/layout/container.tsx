import React, { type FC } from 'react';
import { ScrollView } from 'react-native';

import { makeStyles } from '../../theme';
import { ScreenWrapper } from './screen-wrapper';
import type { ContainerProps } from './types';

const useStyles = makeStyles((theme) => ({
  scrollView: {
    flex: 1,
    backgroundColor: theme.colors.background,
  },
  contentContainer: {
    flexGrow: 1,
  },
}));

export const Container: FC<ContainerProps> = ({
  children,
  style,
  scrollViewProps = {},
  keyboardShouldPersistTaps = 'handled',
}) => {
  const styles = useStyles();

  return (
    <ScreenWrapper style={style}>
      <ScrollView
        style={[styles.scrollView, style]}
        contentContainerStyle={styles.contentContainer}
        keyboardShouldPersistTaps={keyboardShouldPersistTaps}
        showsVerticalScrollIndicator={false}
        {...scrollViewProps}
      >
        {children}
      </ScrollView>
    </ScreenWrapper>
  );
};
