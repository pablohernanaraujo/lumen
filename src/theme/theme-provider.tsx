import React, {
  createContext,
  type FC,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';
import { useColorScheme } from 'react-native';

import { colorTokens } from './colors';
import { baseTokens } from './tokens';
import type {
  Mode,
  Theme,
  ThemeContextValue,
  ThemeProviderProps,
} from './types';

const ThemeContext = createContext<ThemeContextValue | null>(null);

export const ThemeProvider: FC<ThemeProviderProps> = ({
  children,
  preference,
  persistenceKey = '@lumen-theme',
}) => {
  const systemMode = useColorScheme() as Mode | null;
  const [userMode, setUserMode] = useState<Mode | null>(preference || null);

  const currentMode: Mode = userMode ?? systemMode ?? 'light';

  const theme = useMemo<Theme>(
    () => ({
      ...baseTokens,
      colors: colorTokens[currentMode],
      mode: currentMode,
    }),
    [currentMode],
  );

  const setMode = useCallback((mode: Mode): void => {
    setUserMode(mode);
  }, []);

  const toggleMode = useCallback((): void => {
    setUserMode(currentMode === 'light' ? 'dark' : 'light');
  }, [currentMode]);

  const contextValue = useMemo<ThemeContextValue>(
    () => ({
      theme,
      mode: currentMode,
      setMode,
      toggleMode,
    }),
    [theme, currentMode, setMode, toggleMode],
  );

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeContextValue => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
