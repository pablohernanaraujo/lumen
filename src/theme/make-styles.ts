import { StyleSheet } from 'react-native';

import { useTheme } from './theme-provider';
import type { NamedStyles, StylesOrFunction, Theme } from './types';

export const makeStyles =
  <T extends NamedStyles<T>>(
    factory: StylesOrFunction<T>,
  ): (() => StyleSheet.NamedStyles<T>) =>
  (): StyleSheet.NamedStyles<T> => {
    const { theme } = useTheme();

    const styles =
      typeof factory === 'function'
        ? (factory as (theme: Theme) => T)(theme)
        : factory;

    return StyleSheet.create(styles);
  };
