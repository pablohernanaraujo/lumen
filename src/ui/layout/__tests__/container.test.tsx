import React from 'react';
import { Text } from 'react-native';

import { createDualSnapshots, render } from '../../../test-utils';
import { makeStyles } from '../../../theme';
import { Container } from '../container';

const useTestStyles = makeStyles((theme) => ({
  redBackground: {
    backgroundColor: theme.colors.error.main,
  },
}));

describe('Container', () => {
  const mockChildren = <Text>Test Content</Text>;

  createDualSnapshots(<Container>{mockChildren}</Container>);

  describe('with different props', () => {
    it('should render correctly with custom style', () => {
      const TestComponent: React.FC = () => {
        const styles = useTestStyles();
        return (
          <Container style={styles.redBackground}>{mockChildren}</Container>
        );
      };
      const { toJSON } = render(<TestComponent />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render correctly with keyboardShouldPersistTaps set to always', () => {
      const { toJSON } = render(
        <Container keyboardShouldPersistTaps="always">
          {mockChildren}
        </Container>,
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render correctly with keyboardShouldPersistTaps set to never', () => {
      const { toJSON } = render(
        <Container keyboardShouldPersistTaps="never">{mockChildren}</Container>,
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render correctly with custom scrollViewProps', () => {
      const { toJSON } = render(
        <Container
          scrollViewProps={{
            showsHorizontalScrollIndicator: true,
            bounces: false,
            contentInsetAdjustmentBehavior: 'automatic',
          }}
        >
          {mockChildren}
        </Container>,
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render correctly with multiple children', () => {
      const { toJSON } = render(
        <Container>
          <Text>First Child</Text>
          <Text>Second Child</Text>
          <Text>Third Child</Text>
        </Container>,
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('edge cases', () => {
    it('should render correctly with no children', () => {
      const { toJSON } = render(<Container>{null}</Container>);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render correctly with empty scrollViewProps', () => {
      const { toJSON } = render(
        <Container scrollViewProps={{}}>{mockChildren}</Container>,
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
