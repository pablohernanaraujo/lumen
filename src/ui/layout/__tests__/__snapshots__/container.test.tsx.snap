// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container edge cases should render correctly with empty scrollViewProps 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container edge cases should render correctly with no children 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`Container snapshots should render correctly in dark mode 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#000000",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#000000",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container snapshots should render correctly in light mode 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container with different props should render correctly with custom scrollViewProps 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    bounces={false}
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    contentInsetAdjustmentBehavior="automatic"
    keyboardShouldPersistTaps="handled"
    showsHorizontalScrollIndicator={true}
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container with different props should render correctly with custom style 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      {
        "backgroundColor": "#EF4444",
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        {
          "backgroundColor": "#EF4444",
        },
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container with different props should render correctly with keyboardShouldPersistTaps set to always 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="always"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container with different props should render correctly with keyboardShouldPersistTaps set to never 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="never"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Test Content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Container with different props should render correctly with multiple children 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View>
      <Text>
        First Child
      </Text>
      <Text>
        Second Child
      </Text>
      <Text>
        Third Child
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
